### PR TITLE
Handle more cases in both new line formatting and curly brace completion

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes https://github.com/dotnet/roslyn/issues/50275

TODO:

- [x] Tests
- [x] Discuss whether `CurlyBraceCompletionService` should respect other new lines options for braces. Currently it [only respects `NewLinesForBracesInObjectCollectionArrayInitializers`](https://github.com/Youssef1313/roslyn/blob/15b1443badf2bb2d2491a93c01f77e59a0962639/src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs#L444-L447). (@sharwell @CyrusNajmabadi What are your thoughts on this point?)
- [x] Try to share some common code between `CurlyBraceCompletionService` and `NewLineUserSettingFormattingRule`. (I'm not sure if it's worth it, nor where the common code should exist - so dropping this for now)

### CODE REVIEW

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Yeah I'll add them. Wanted to first make sure if I should extend curly brace completion (details in PR description) and will add all the needed tests

**Comment by Youssef1313** [comment]
@CyrusNajmabadi I added some tests. Let me know if you want to add more.

File: src/EditorFeatures/CSharpTest/AutomaticCompletion/AutomaticBraceCompletionTests.cs [hunk 1]
```diff
@@ -1348,7 +1342,506 @@ public void CurlyBraceFormatting_InsertsCorrectNewLine()
...
+            var expected = @"
+class C
+{
+    public void man()
+    {
+        if (x is string {
+
+        }
+    }
+}";
```
Comment (Youssef1313) [inline src/EditorFeatures/CSharpTest/AutomaticCompletion/AutomaticBraceCompletionTests.cs:1436]: I'm feeling the indentation is unexpected here.

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Can you review please?

File: src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs [hunk 2]
```diff
@@ -380,20 +385,110 @@ public override AbstractFormattingRule WithOptions(AnalyzerConfigOptions options
...
+                if (needsNewLine == true)
+                {
+                    return CreateAdjustNewLinesOperation(1, AdjustNewLinesOption.PreserveLines);
+                }
+                else if (needsNewLine == false)
+                {
+                    return null;
                 }
 
                 return base.GetAdjustNewLinesOperation(in previousToken, in currentToken, in nextOperation);
```
Comment (sharwell) [inline src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:494]: 💡 This seems like a good use of a switch expression

File: src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs [hunk 3]
```diff
@@ -370,8 +370,13 @@ public override AbstractFormattingRule WithOptions(AnalyzerConfigOptions options
                 return new BraceCompletionFormattingRule(_indentStyle, cachedOptions);
             }
 
-            public override AdjustNewLinesOperation? GetAdjustNewLinesOperation(in SyntaxToken previousToken, in SyntaxToken currentToken, in NextGetAdjustNewLinesOperation nextOperation)
+            private static bool? NeedsNewLine(in SyntaxToken currentToken, CachedOptions options)
```
Comment (sharwell) [inline src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs:373]:
```suggestion
            private static bool? NeedsNewLine(in SyntaxToken currentToken, in CachedOptions options)
```

File: src/EditorFeatures/CSharpTest/AutomaticCompletion/AutomaticBraceCompletionTests.cs [hunk 4]
```diff
@@ -1348,7 +1342,506 @@ public void CurlyBraceFormatting_InsertsCorrectNewLine()
             CheckReturn(session.Session, 4, result: "class C\r{\r\r}");
         }
 
-        internal static Holder CreateSession(string code, Dictionary<OptionKey2, object> optionSet = null)
+        [WorkItem(50275, "https://github.com/dotnet/roslyn/issues/50275")]
+        [WpfFact, Trait(Traits.Feature, Traits.Features.AutomaticCompletion)]
+        public void WithInitializer_OpenBraceOnSameLine_Enter()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/AutomaticCompletion/AutomaticBraceCompletionTests.cs:1347]: 💡 This test can cover both cases. Start by defining it as a theory:

```suggestion
        [WorkItem(50275, "https://github.com/dotnet/roslyn/issues/50275")]
        [WpfTheory, CombinatorialData, Trait(Traits.Feature, Traits.Features.AutomaticCompletion)]
        public void WithInitializer_Enter(bool bracesOnNewLine)
```

Then create locals for both `expectedForSameLine` and `expectedForNewLine` and set the option and assertion according to the argument and correct local, respectively. This test form is not only condensed, but it more clearly shows the impact of the distinct option values considering the starting `code` is known to be identical for both tests.

This suggestion applies for each pair added in this PR.

File: src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs (hunk 3 above)
Comment (Youssef1313) [reply src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs]: @sharwell In a recent refactoring, `CachedOptions` was completely removed and a new class (as opposed to struct) is introduced. So I think no action is needed for this comment now.

File: src/EditorFeatures/CSharpTest/AutomaticCompletion/AutomaticBraceCompletionTests.cs (hunk 4 above)
Comment (Youssef1313) [reply src/EditorFeatures/CSharpTest/AutomaticCompletion/AutomaticBraceCompletionTests.cs]: @sharwell Thanks! Fixed.

**Review by sharwell — APPROVED** [review]

**Pull request merged**